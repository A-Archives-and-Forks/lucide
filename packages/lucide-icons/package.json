{
  "name": "@lucide/icons",
  "description": "A Lucide icon library that contains icon data in a standard Lucide format.",
  "version": "0.0.1",
  "license": "ISC",
  "homepage": "https://lucide.dev",
  "bugs": "https://github.com/lucide-icons/lucide/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/lucide-icons/lucide.git",
    "directory": "packages/lucide"
  },
  "keywords": [
    "Lucide",
    "HTML",
    "Feather",
    "Icons",
    "Icon",
    "SVG",
    "Feather Icons"
  ],
  "amdName": "lucide-icons",
  "source": "src/lucide-icons.ts",
  "main": "dist/cjs/lucide-icons.cjs",
  "module": "dist/esm/lucide-icons.mjs",
  "types": "dist/esm/lucide-icons.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/esm/lucide-icons.d.ts",
      "import": "./dist/esm/lucide-icons.mjs",
      "require": "./dist/cjs/lucide-icons.cjs",
      "default": "./dist/esm/lucide-icons.mjs"
    },
    "./icons/*": {
      "import": "./dist/esm/icons/*.mjs",
      "default": "./dist/esm/icons/*.mjs"
    },
    "./dynamic": {
      "types": "./dist/esm/dynamic.d.ts",
      "import": "./dist/esm/dynamic.mjs",
      "require": "./dist/cjs/dynamic.cjs",
      "default": "./dist/esm/dynamic.mjs"
    },
    "./build": {
      "types": "./dist/esm/build.d.ts",
      "import": "./dist/esm/build.mjs",
      "require": "./dist/cjs/build.cjs",
      "default": "./dist/esm/build.mjs"
    }
  },
  "scripts": {
    "build": "pnpm clean && pnpm copy:license && pnpm build:icons && pnpm build:bundle",
    "copy:license": "cp ../../LICENSE ./LICENSE",
    "clean": "rm -rf dist && rm -rf stats && rm -rf ./src/icons/*.ts",
    "build:icons": "build-icons --output=./src --templateSrc=./scripts/exportTemplate.mjs --renderUniqueKey --withAliases --withDynamicImports --separateAliasesFile --aliasesFileExtension=.ts --iconFileExtension=.ts --exportFileName=index.ts",
    "build:bundle": "rollup -c rollup.config.mjs",
    "test": "pnpm build:icons && vitest run",
    "test:watch": "vitest watch",
    "test:update": "vitest -u",
    "version": "pnpm version --git-tag-version=false"
  },
  "devDependencies": {
    "@lucide/build-icons": "workspace:*",
    "@lucide/helpers": "^1.0.0",
    "@lucide/rollup-plugins": "workspace:*",
    "@rollup/plugin-replace": "^6.0.2",
    "@testing-library/jest-dom": "^6.6.3",
    "jest-serializer-html": "^7.1.0",
    "rollup": "^4.40.0",
    "rollup-plugin-dts": "^6.2.1",
    "rollup-plugin-preserve-directives": "^0.4.0",
    "typescript": "^4.9.5",
    "vite": "5.4.14",
    "vitest": "^1.6.1"
  }
}
