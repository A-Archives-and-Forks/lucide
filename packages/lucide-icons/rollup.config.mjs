import plugins from '@lucide/rollup-plugins';
import pkg from './package.json' with {type: 'json'};
import dts from 'rollup-plugin-dts';

const packageName = '@lucide/icons';
const outputFileName = 'lucide-icons';
const inputs = [`src/lucide-icons.ts`];
const bundles = [
  {
    format: 'umd',
    inputs,
    minify: true,
  },
  {
    format: 'umd',
    inputs,
  },
  {
    format: 'cjs',
    inputs,
    extension: 'cjs',
  },
  {
    format: 'esm',
    inputs: [...inputs, './src/dynamic.ts', './src/build.ts'],
    preserveModules: true,
    extension: 'mjs',
  },
  /*
  {
    format: 'esm',
    inputs: ['src/dynamic.ts'],
    outputFileName: 'dynamic',
    external: [/src/],
    paths: (id) => {
      if (id.match(/src/)) {
        const [, modulePath] = id.match(/src\/(.*)\.ts/);

        return `${modulePath}.js`;
      }
    },
  },
  {
    format: 'esm',
    inputs: ['src/build.ts'],
    outputFileName: 'build.mjs',
    external: [/src/],
    paths: (id) => {
      if (id.match(/src/)) {
        const [, modulePath] = id.match(/src\/(.*)\.ts/);

        return `${modulePath}.js`;
      }
    },
  },
   */
];

const configs = bundles
  .map(
    ({
       inputs,
       outputDir = 'dist',
       outputFile,
       format,
       minify,
       preserveModules,
       entryFileNames,
       external = [],
       paths,
       extension = 'js',
     }) =>
      inputs.map((input) => ({
        input,
        plugins: [
          ...plugins({pkg, minify}),
        ],
        external,
        output: {
          name: packageName,
          entryFileNames,
          ...(preserveModules
            ? {
              dir: `${outputDir}/${format}`,
              entryFileNames: `[name].${extension}`,
            }
            : {
              file: outputFile ?? `${outputDir}/${format}/${outputFileName}${minify ? '.min' : ''}.${extension}`,
            }),
          paths,
          format,
          sourcemap: true,
          preserveModules,
          exports: 'named',
          globals: {},
          preserveModulesRoot: 'src',
        },
      })),
  )
  .flat();

export default [
  ...[
    outputFileName,
    `${outputFileName}.prefixed`,
    `${outputFileName}.suffixed`,
    'dynamic',
    'build'
  ].map(filename => (
    {
      input: `./src/${filename}.ts`,
      output: [{
        file: `dist/esm/${filename}.d.ts`, format: 'esm'
      }, {
        file: `dist/cjs/${filename}.d.cts`, format: 'cjs'
      }, {
        file: `dist/umd/${filename}.d.ts`, format: 'umd'
      }],
      plugins: [dts()],
    })
  ),
  ...configs,
];
