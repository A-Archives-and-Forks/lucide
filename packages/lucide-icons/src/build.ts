export { default as buildLucideIconNode } from './buildLucideIconNode';
export { default as buildLucideDataUri } from './buildLucideDataUri';
export { default as buildLucideIconElement } from './buildLucideIconElement';
export { default as buildLucideSvg } from './buildLucideSvg';
