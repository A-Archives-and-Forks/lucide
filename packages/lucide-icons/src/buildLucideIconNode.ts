import { LucideBuildParams, LucideIcon, LucideIconNode } from './types';

const defaultAttributes = {
  xmlns: 'http://www.w3.org/2000/svg',
  width: '24',
  height: '24',
  viewBox: '0 0 24 24',
  fill: 'none',
  stroke: 'currentColor',
  'stroke-width': '2',
  'stroke-linecap': 'round',
  'stroke-linejoin': 'round',
};

function buildLucideIconNode(icon: LucideIcon, params: LucideBuildParams = {}): LucideIconNode {
  const viewBoxWidth = ('size' in icon ? icon.size : icon.width) ?? defaultAttributes.width;
  const viewBoxHeight = ('size' in icon ? icon.size : icon.height) ?? defaultAttributes.height;
  const attributes = {
    ...defaultAttributes,
    ...('color' in params && { stroke: params['color'] }),
    ...('size' in params &&
      params['size'] && {
        width: params['size'].toString(10),
        height: params['size'].toString(10),
      }),
    ...('width' in params && params['width'] && { width: params['width'].toString(10) }),
    ...('height' in params && params['height'] && { height: params['height'].toString(10) }),
    ...('strokeWidth' in params &&
      params['strokeWidth'] && { 'stroke-width': params['strokeWidth'].toString(10) }),
    ...('absoluteStrokeWidth' in params &&
      params['absoluteStrokeWidth'] && { 'vector-effect': 'non-scaling-stroke' }),
    class: `lucide lucide-${icon.name} ${params['className'] ?? ''}`.trim(),
    viewBox: `0 0 ${viewBoxWidth} ${viewBoxHeight}`,
    ...('attributes' in params && params.attributes),
  };
  return ['svg', attributes, icon.node];
}

export default buildLucideIconNode;
