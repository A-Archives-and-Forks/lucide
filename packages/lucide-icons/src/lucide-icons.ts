export * as icons from './icons';
export * from './icons';
export * from './aliases/index';
export * from './types';
