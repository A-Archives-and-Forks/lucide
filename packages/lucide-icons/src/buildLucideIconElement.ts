import { LucideBuildParams, LucideIcon, LucideIconNode } from './types';
import buildLucideIconNode from './buildLucideIconNode';

const buildDomElement = (
  document: Document,
  [tagName, attributes, children = []]: LucideIconNode,
): Element => {
  const element = document.createElementNS('http://www.w3.org/2000/svg', tagName);
  for (const [attrName, value] of Object.entries(attributes)) {
    element.setAttribute(attrName, value);
  }
  for (const node of children) {
    const childNode = buildDomElement(document, node);
    element.appendChild(childNode);
  }
  return element;
};

function buildLucideIconElement(
  document: Document,
  icon: LucideIcon,
  params: LucideBuildParams = {},
) {
  return buildDomElement(document, buildLucideIconNode(icon, params));
}

export default buildLucideIconElement;
