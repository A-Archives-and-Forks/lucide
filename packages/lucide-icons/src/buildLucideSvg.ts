import { LucideBuildParams, LucideIcon, LucideIconNode } from './types';
import buildLucideIconNode from './buildLucideIconNode';

const buildDomNode = ([tagName, attributes, children = []]: LucideIconNode): string =>
  `<${tagName} ${Object.entries(attributes)
    .map(([attrName, value]) => `${attrName}="${value}"`)
    .join(' ')}>${children?.map((child) => buildDomNode(child)).join('')}</${tagName}>`;

function buildLucideSvg(icon: LucideIcon, params: LucideBuildParams = {}) {
  return buildDomNode(buildLucideIconNode(icon, params));
}

export default buildLucideSvg;
