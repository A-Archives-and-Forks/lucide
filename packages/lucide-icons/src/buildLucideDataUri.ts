import { LucideBuildParams, LucideIcon } from './types';
import buildLucideSvg from './buildLucideSvg';

function buildLucideDataUri(icon: LucideIcon, params: LucideBuildParams = {}) {
  return `data:image/svg+xml;base64,${btoa(buildLucideSvg(icon, params))}`;
}

export default buildLucideDataUri;
