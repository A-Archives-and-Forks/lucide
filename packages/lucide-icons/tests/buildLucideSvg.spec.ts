import { describe, expect, it } from 'vitest';
import { House } from '../src/lucide-icons';
import { getOriginalSvg, removeKeys } from './helpers';
import buildLucideSvg from '../src/buildLucideSvg';
import buildLucideDataUri from '../src/buildLucideDataUri';

describe('buildLucideSvg', () => {
  it('should match the snapshot', () => {
    const HouseSVG = buildLucideSvg(House);

    expect(HouseSVG).toMatchSnapshot();
  });

  it('should create the correct svg element', () => {
    const HouseSVG = buildLucideSvg(House);

    const svg = getOriginalSvg('house', undefined, true);

    expect(removeKeys(HouseSVG)).toBe(svg);
  });
});

describe('buildLucideSvgDataUri', () => {
  it('should match the snapshot', () => {
    const HouseDataUri = buildLucideDataUri(House);

    expect(HouseDataUri).toMatchSnapshot();
  });
});
