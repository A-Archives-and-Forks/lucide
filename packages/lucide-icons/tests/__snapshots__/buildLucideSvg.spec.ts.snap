// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildLucideSvg > should match the snapshot 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="24"
     height="24"
     viewbox="0 0 24 24"
     fill="none"
     stroke="currentColor"
     stroke-width="2"
     stroke-linecap="round"
     stroke-linejoin="round"
     class="lucide lucide-house"
>
  <path d="M15 21v-8a1 1 0 0 0-1-1h-4a1 1 0 0 0-1 1v8"
        key="5wwlr5"
  >
  </path>
  <path d="M3 10a2 2 0 0 1 .709-1.528l7-5.999a2 2 0 0 1 2.582 0l7 5.999A2 2 0 0 1 21 10v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
        key="1d0kgt"
  >
  </path>
</svg>
`;

exports[`buildLucideSvgDataUri > should match the snapshot 1`] = `"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgdmlld0JveD0iMCAwIDI0IDI0IiBmaWxsPSJub25lIiBzdHJva2U9ImN1cnJlbnRDb2xvciIgc3Ryb2tlLXdpZHRoPSIyIiBzdHJva2UtbGluZWNhcD0icm91bmQiIHN0cm9rZS1saW5lam9pbj0icm91bmQiIGNsYXNzPSJsdWNpZGUgbHVjaWRlLWhvdXNlIj48cGF0aCBkPSJNMTUgMjF2LThhMSAxIDAgMCAwLTEtMWgtNGExIDEgMCAwIDAtMSAxdjgiIGtleT0iNXd3bHI1Ij48L3BhdGg+PHBhdGggZD0iTTMgMTBhMiAyIDAgMCAxIC43MDktMS41MjhsNy01Ljk5OWEyIDIgMCAwIDEgMi41ODIgMGw3IDUuOTk5QTIgMiAwIDAgMSAyMSAxMHY5YTIgMiAwIDAgMS0yIDJINWEyIDIgMCAwIDEtMi0yeiIga2V5PSIxZDBrZ3QiPjwvcGF0aD48L3N2Zz4="`;
